src/util/IotaIterator.hpp

```diff
@@ -0,0 +1,233 @@
+
+    /**
+     * @brief Construct a new Iota Iterator object
+     * @param v The value of the iterator
+     */
+    IotaIterator(int v) : value_(v)
+    {
+    }
+
+    IotaIterator(IotaIterator const&) = default;
+
+    IotaIterator&
+    operator=(IotaIterator const&) = default;
+
+    /**
+     * @brief Dereference operator, return the value_
+     * @return int const& The value_'s const reference
+     */
+    int const&
+    operator*() const noexcept
+    {
+        return value_;
+    }
+
+    /**
+     * @brief += operator, increment the value_ by offset
+     * @param offset The offset to increment the value_
+     * @return IotaIterator& The iterator itself
+     */
+    IotaIterator&
+    operator+=(int offset) noexcept
+    {
+        value_ += offset;
+        return *this;
+    }
+
+    /**
+     * @brief -= operator, decrement the value_ by offset
+     * @param offset The offset to decrement the value_
+     * @return IotaIterator& The iterator itself
+     */
+    IotaIterator&
+    operator-=(int offset) noexcept
+    {
+        value_ += (-offset);
+        return *this;
+    }
+
+    /**
+     * @brief ++ operator, increment the value_ by 1 and return the iterator itself
+     * @return IotaIterator The iterator itself
+     */
+    IotaIterator
+    operator++() noexcept
+    {
+        ++value_;
+        return *this;
+    }
+
+    /**
+     * @brief ++(int) operator, return a new iterator of value_ then increment itself value_ by 1
+     * @return IotaIterator The new iterator
+     */
+    IotaIterator
+    operator++(int) noexcept
+    {
+        IotaIterator tmp(*this);
+        ++(*this);
+        return tmp;
+    }
+
+    /**
+     * @brief -- operator, decrement the value_ by 1 and return the iterator itself
+     * @return IotaIterator The iterator itself
+     */
+    IotaIterator
+    operator--() noexcept
+    {
+        --value_;
+        return *this;
+    }
+
+    /**
+     * @brief --(int) operator, return a new iterator of value_ then decrement itself value_ by 1
+     * @return IotaIterator The new iterator
+     */
+    IotaIterator
+    operator--(int) noexcept
+    {
+        IotaIterator tmp(*this);
+        --(*this);
+        return tmp;
+    }
+
+    /**
+     * @brief offset operator, return the iterator that is offset away from the current iterator
+     * @param offset The offset to move the iterator
+     * @return IotaIterator The new iterator that is offset away from the current iterator
+     */
+    IotaIterator
+    operator[](int offset) const noexcept
+    {
+        return IotaIterator(value_ + offset);
+    }
+};
+
+/**
+ * @brief Compare two IotaIterator objects, return true if they contains the same value_
+ * @param lhs The left hand side IotaIterator object
+ * @param rhs The right hand side IotaIterator object
+ * @return true If they contains the same value_
+ */
+bool
+operator==(IotaIterator const& lhs, IotaIterator const& rhs) noexcept;
```

Comment:
This and below look like the 'friend' versions. Do we actually need them at all? would it not work with just member functions that take rhs and compare with *this? Otherwise shouldn't they be hidden friends instead?